src/bundles/peer-locations.js

```diff
@@ -12,243 +13,90 @@ export default function (opts) {
   // Cache options
   opts.cache = opts.cache || {}
 
-  const defaultState = {
-    // Peer locations keyed by peer ID then peer address.
-    // i.e. { [peerId]: { [multiaddr]: { state, data?, error? } } }
-    // `state` can be queued, resolving, resolved or failed
-    // `data` is the resolved location data (see ipfs-geoip docs)
-    // `error` is only present if state is 'failed'
-    locations: {},
-    // Peer IDs in the queue for resolving ONE of their queued addresses.
-    // We actually have a queue of queues since each peer may have multiple
-    // addresses to resolve. Peer locations are resolved in PARALLEL but for
-    // any given peer we work our way through it's list of addresses in SERIES
-    // until we find one that resolves.
-    queuingPeers: [],
-    // Peer IDs currently resolving for one of their queued addresses
-    resolvingPeers: []
-  }
-
   const peerLocResolver = new PeerLocationResolver(opts)
 
-  return {
+  const bundle = createAsyncResourceBundle({
     name: 'peerLocations',
-
-    reducer (state = defaultState, action) {
-      if (action.type === 'PEER_LOCATIONS_PEERS_QUEUED') {
-        const { addrsByPeer, peerByAddr } = action.payload
-
-        return {
-          ...state,
-          queuingPeers: state.queuingPeers.concat(Object.keys(addrsByPeer)),
-          locations: Object.keys(peerByAddr).reduce((locs, addr) => {
-            const peerId = peerByAddr[addr]
-            locs[peerId] = locs[peerId] || {}
-            locs[peerId][addr] = { state: 'queued' }
-            return locs
-          }, { ...state.locations })
-        }
-      }
-
-      if (action.type === 'PEER_LOCATIONS_RESOLVE_STARTED') {
-        const { peerId, addr } = action.payload
-
-        return {
-          ...state,
-          queuingPeers: state.queuingPeers.filter(id => id !== peerId),
-          resolvingPeers: state.resolvingPeers.concat(peerId),
-          locations: {
-            ...state.locations,
-            [peerId]: {
-              ...state.locations[peerId],
-              [addr]: { state: 'resolving' }
-            }
+    actionBaseType: 'PEER_LOCATIONS',
+    getPromise: async ({ store, getIpfs }) => {
+      const peers = store.selectPeers()
+      const locations = store.selectPeerLocations() || {}
+
+      for (const p of peers) {
+        const peerId = p.peer.toB58String()
+        const addr = p.addr.toString()
+
+        if (!locations[peerId]) {
+          const res = peerLocResolver.checkPeer(peerId, addr, getIpfs)
```

Comment:
Let's see if we can find a more intutive contract for checkPeer... it either returns a location synchronously or queues up some async work. It might be a little more intuitive if we pass the list of peers to PeerLocationsResolver, and await list of locations it can find in its cache... so the contract of the function goes from being sync or null with some async work to always async, with some async taking longer than others...

```js
const locations = await PeerLocationResolver.findLocations(peers, getIpfs)
```

It's basically moving the loop, but I'd have less async to juggle in my head.

Reply:
Thanks for the suggestion! Just implemented!